### PR TITLE
Add relevance metrics including pruned tokens to MS Marco ranking track

### PR DESCRIPTION
**Pull request by kderusso**
Adds recall metrics for query pruning using the [weighted tokens](https://www.elastic.co/guide/en/elasticsearch/reference/master/query-dsl-weighted-tokens-query.html) query and precomputed ELSER tokens.

### CODE REVIEW

File: msmarco-passage-ranking/operations/default.json [hunk 1]
```diff
@@ -1,70 +1,152 @@
...
+   "name": "text-expansion-splade-search-maxwand-enabled",
+   "operation-type": "search",
+   "param-source": "query_param_source",
+   "query_source": "queries.json",
+   "query_strategy": "text_expansion",
+   "text_expansion_field": "text_expansion_splade",
+   "track_total_hits": false
+},
+{
+   "name": "pruned-text-expansion-search-maxwand-disabled",
```
Comment (kderusso) [inline msmarco-passage-ranking/operations/default.json:38]: Added operations here

File: msmarco-passage-ranking/operations/default.json
Comment (kderusso) [inline msmarco-passage-ranking/operations/default.json:1]: This file was auto-formatted for correct json indentation. I have left comments starting at the two places where I made changes.

File: msmarco-passage-ranking/track.py [hunk 2]
```diff
@@ -80,5 +184,128 @@ def params(self):
...
+        self._qrels = read_qrels(os.path.join(cwd, self._qrels_file))
+
+    def partition(self, partition_index,
+                  total_partitions):
+        return self
+
+    def params(self):
+        return {
+            "index": self._index_name,
+            "cache": self._params.get("cache", False),
```
Comment (demjened) [inline msmarco-passage-ranking/track.py:219]: Any reason for not using `self._cache` here?

File: msmarco-passage-ranking/track.py [hunk 3]
```diff
@@ -80,5 +184,128 @@ def params(self):
         }
 
 
+# For each query this will generate the weighted terms query, a pruned version and a rescored pruned version of the same query.
```
Comment (demjened) [inline msmarco-passage-ranking/track.py:187]: Looks like the generation logic is in `WeightedTermsRecallRunner` - consider moving the comment.

**Review by demjened — APPROVED** [review]
LGTM with a few minor comments

File: msmarco-passage-ranking/README.md
Comment (Mikep86) [inline msmarco-passage-ranking/README.md:1]: Can you document the new parameters in the README?

**Review by Mikep86 — COMMENTED** [review]
Looking good! Left a few minor comments.

File: msmarco-passage-ranking/challenges/default.json [hunk 4]
```diff
@@ -70,14 +70,40 @@
...
+    },
     {
       "operation": "text-expansion-splade-search-maxwand-disabled",
       "warmup-iterations": 100,
       "iterations": 1000,
       "clients": {{search_clients | default(1)}}
-  },
+    },
+    {
+      "operation": "pruned-text-expansion-search-maxwand-enabled",
```
Comment (saikatsarkar056) [inline msmarco-passage-ranking/challenges/default.json:82]: I suggest adding a "name" field besides "operation" field. This is helpful if you want to include/exclude a specific task while running benchmark. [This](https://github.com/elastic/rally-tracks/blob/master/wikipedia/challenges/default.json#L41-L42) is an example.

File: elastic/Makefile [hunk 5]
```diff
@@ -59,7 +59,7 @@ install: venv-create
 	# install pytest for tests
 	. $(VENV_ACTIVATE_FILE); pip3 install pytest==6.2.5 pytest-benchmark==3.2.2
 	# install dependencies for tests
-	. $(VENV_ACTIVATE_FILE); pip3 install geneve==0.0.3 pytest-asyncio==0.18.1 git+https://github.com/elastic/package-assets.git
+	. $(VENV_ACTIVATE_FILE); pip3 install geneve==0.0.3 pytest-asyncio==0.18.1 pytrec_eval numpy git+https://github.com/elastic/package-assets.git
```
[the remark this answers is not in the document]
Comment (afoucret) [reply elastic/Makefile:62]: Also suggest to pin versions here.

File: msmarco-passage-ranking/track.json [hunk 6]
```diff
@@ -2,6 +2,10 @@
 {
   "version": 2,
   "description": "Benchmark bm25, semantic and hybrid search on the MS MARCO passage dataset",
+  "dependencies": [
+    "pytrec_eval",
+    "numpy"
+  ],
```
[the remark this answers is not in the document]
Comment (afoucret) [reply msmarco-passage-ranking/track.json:8]: Not to my knowledge which is very low on this specific topic.

[the remark this answers is not in the document]
Comment (gbanasiak) [reply msmarco-passage-ranking/track.json:8]: The `dependencies` section in `track.json` is the right approach although external dependencies are not properly handled by track listing which we will try to address in https://github.com/elastic/rally/pull/1817. In the meantime, please try importing `pytrec_eval` late in `track.py`. Please use pinned version as documented in https://esrally.readthedocs.io/en/stable/track.html?highlight=dependencies#dependencies.

File: pyproject.toml [hunk 7]
```diff
@@ -40,6 +42,8 @@ dependencies = [
 [tool.hatch.envs.unit]
 extra-dependencies = [
     "geneve==0.0.3",
+    "pytrec_eval",
+    "numpy",
```
[the remark this answers is not in the document]
Comment (gbanasiak) [reply pyproject.toml:46]: That's not needed. These dependencies are only meaningful when running unit tests, like the ones in `elastic/tests`. There are no unit tests in `msmarco-passage-ranking` track.

File: pyproject.toml [hunk 8]
```diff
@@ -10,6 +10,8 @@ develop = [
     "isort==5.12.0",
     "pre-commit==3.3.3",
     "pip==22.2",
+    "pytrec_eval",
+    "numpy",
```
[the remark this answers is not in the document]
Comment (gbanasiak) [reply pyproject.toml:14]: Not needed, see earlier comment about `dependencies` section.

File: msmarco-passage-ranking/track.py [hunk 9]
```diff
@@ -1,5 +1,127 @@
...
 import os
+from collections import defaultdict
+
+import pytrec_eval as pe
+
+Qrels = dict[str, dict[str, int]]
+Results = dict[str, dict[str, float]]
+
+
+def calc_ndcg(qrels: Qrels, results: Results, k_list: list) -> dict:
```
Comment (gbanasiak) [inline msmarco-passage-ranking/track.py:13]: Please try moving `pytrec_eval` import into `calc_ndcg()`.

File: msmarco-passage-ranking/track.py [hunk 10]
```diff
@@ -1,5 +1,127 @@
...
+                ]
+            }
+        }
+    }
+
+
+def generate_pruned_query(field, query_expansion, boost=1.0):
+    return {
+        "query": {
+            "weighted_tokens": {
```
Comment (gbanasiak) [inline msmarco-passage-ranking/track.py:87]: I've tried running the modified track and noticed `weighted_tokens` query is not available in 8.11.x. Is this planned for 8.13.0? We need to apply the right versioning as per https://esrally.readthedocs.io/en/stable/track.html#custom-track-repositories. The most recent non-master branch in the track repo is `8.11`. We need to bump up to `8.13` before merging this change.

Comment (kderusso) [reply msmarco-passage-ranking/track.py:87]: Correct, this is for 8.13.0.

File: msmarco-passage-ranking/challenges/default.json (hunk 4 above)
Comment (kderusso) [reply msmarco-passage-ranking/challenges/default.json:82]: Good call, adding!

File: msmarco-passage-ranking/_tools/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,2 @@
+pytrec_eval
+numpy
```
Comment (gbanasiak) [inline msmarco-passage-ranking/_tools/requirements.txt:2]: Should the presence of this file be mentioned in README? Also, please align version pinning with `dependencies` section.

**Review by gbanasiak — APPROVED** [review]
LGTM. Left small remark regarding `_tools/requirements.txt` file.

**Review by Mikep86 — APPROVED** [review]
LGTM!

**Pull request merged**